reps/2023-03-10-batch-inference.md

```diff
@@ -227,28 +237,176 @@ def preprocess(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+1. Introduce a DeveloperAPI `BatchMappable` interface that `Preprocessor` and `Predictor` subclass. This interface will define the following:
+- Logic for the actual computation (either prediction or preprocessing)
+- Validating the user provided batch_format
+- Validating the user provided compute strategy
 
+```python
+@DeveloperAPI
+class BatchMappable(Callable[[T], U]):
+   """Subclass this if you want your class to be passable to Dataset.map_batches."""
+   
+    def __call__(self, batch: T) -> U:
+        """Transform the provided batch.
+
+        Args:
+            batch: The batch to transform.
+        """
+        
+        raise NotImplementedError
+
+    def setup(self):
+        """Additional setup logic to run.
+
+        If a callable class is being used, this function will be run in the `__init__`
+        of the Callable class.
+        """
+        pass
+       
+    def validate_batch_format(self, batch_format: Optional[str]) -> Optional[str]:
+        """Validate that the provided `batch_format` is compatible with this `BatchMappable`.
+
+        Raises an error if the provided `batch_format` is not supported with this `BatchMappable` 
+        (for example Pyarrow format for Predictors.)
+
+        If `batch_format` is "default", returns the preferred `batch_format` for this `BatchMappable`.
+
+        Args:
+            batch_format: The `batch_format` that is provided to `map_batches`.
+
+        Returns:
+            The batch format to use.
+        """
+       return None
+
+    def validate_compute_strategy(self, strategy):
+        """Validate that the provided `strategy` is compatible with this `BatchMappable`.
+
+        For example, enforce that `ActorPoolStrategy` is provided for Predictors.
+        """
+        pass
+
+    def use_callable_class(self) -> bool:
+        """Indicates whether this `BatchMappable` should use a callable class."""
+```
+
+`map_batches` will support instances of `BatchMappable` as a supported UDF
+
+2. Add a private `to_checkpoint` serializer for `Predictor`, analagous to the current `from_checkpoint` deserializer. 
+
+```python
+class Predictor:
+    def _to_checkpoint(self) -> Checkpoint:
+        """Create a serializable Checkpoint from this Predictor.
+
+        Returns:
+            Checkpoint containing the state for this Predictor.
+        """
+    
+    def __getstate__(self):
+        return self._to_checkpoint()
+
+    def __setstate__(self, checkpoint):
+        # Same logic as `from_checkpoint`.
+
+    def use_callable_class(self) -> bool:
+        return True
+```
+
+3. When using Predictors with AIR Checkpoints, move the construction logic to `setup` so that it can be called within the actor.
+
+```python
+class Predictor:
+    def _to_checkpoint(self) -> Checkpoint:
+        """Create a serializable Checkpoint from this Predictor.
+
+        Returns:
+            Checkpoint containing the state for this Predictor.
+        """
+        ...
+    
+    def setup(self):
```

Comment:
Is the idea this is called on first use of the predictor? Should it be internal?

Reply:
It should never need to be called by users.

But, inside `map_batches` it is called here: https://github.com/ray-project/enhancements/pull/25/files#diff-0b39e135189387c4d60ea6f6d574d05c99eeb634ef29e2abae5a92442ce326fdR388

Reply:
I don't think we need this though right? For non-lazy Predictors, there is no setup. For lazy predictors, you can lazily do the setup on the first call.

Hence, we can eliminate this API.

Reply:
It would be preferable to call this separately here: https://github.com/ray-project/enhancements/pull/25/files#diff-0b39e135189387c4d60ea6f6d574d05c99eeb634ef29e2abae5a92442ce326fdR388

Otherwise, we can't start the initialization until there are actual prediction tasks to dispatch.

This is purely a developer facing API, so has no user impact.

Reply:
I see, that makes sense (create as soon as actor is created). Would be good to add this as a motivation for this API.

Reply:
Clarified in the docstring for `setup`